controller/test/pid_lib/pid_test.cpp

```diff
@@ -0,0 +1,393 @@
+  EXPECT_NEAR(output, integral * Ki, OUTPUT_TOLERANCE);
+  // Advance time
+  uint32_t dt = sample_time;
+  Hal.delay(dt);
+  myPID.Compute();
+  // Expect output to integrate this
+  integral += (setpoint - input) * (dt / 1000.0);
+  EXPECT_NEAR(output, integral * Ki, OUTPUT_TOLERANCE);
+  // Change sample time and advance
+  sample_time = 50;
+  dt = sample_time;
+  myPID.SetSampleTime(sample_time);
+  integral += (setpoint - input) * (dt / 1000.0);
+  Hal.delay(dt);
+  myPID.Compute();
+  EXPECT_NEAR(output, integral * Ki, OUTPUT_TOLERANCE);
+}
+
+TEST(pidTest, MissedSample) {
+  // This test uses integral to check the effect of missing a sample in the
+  // execution of PID
+  float Kp = 0;
+  float Ki = 0.2;
+  float Kd = 0;
+  double output = 0;
+  PID myPID(&input, &output, &setpoint, Kp, Ki, Kd, DIRECT);
+  myPID.SetSampleTime(sample_time);
+  myPID.SetMode(AUTOMATIC);
+  myPID.Compute();
+
+  // First task, with only Ki set, output = error*sample_time
+  double integral = (setpoint - input) * (sample_time / 1000.0);
+  EXPECT_NEAR(output, integral * Ki, OUTPUT_TOLERANCE);
+  // Advance time
+  uint32_t dt = 2 * sample_time;
+  Hal.delay(dt);
+  myPID.Compute();
+  // Expect output to integrate this
+  integral += (setpoint - input) * (dt / 1000.0);
+  EXPECT_NEAR(output, integral * Ki, OUTPUT_TOLERANCE);
+  // check that a new call without change in time has no effect, even if we
+  // missed a sample
+  double last_output = output;
+  myPID.Compute();
+  EXPECT_EQ(output, last_output);
+  // Advance time a small amount to allow PID to catch up to missed sample
+  dt = max_task_jitter;
+  Hal.delay(dt);
+  integral += (setpoint - input) * (dt / 1000.0);
+  myPID.Compute();
+  // Expect output to have a new update
+  EXPECT_NEAR(output, integral * Ki, OUTPUT_TOLERANCE);
+}
+
+TEST(pidTest, OffOnCycle) {
+  // This test switches the PID and checks the output
+  float Kp = 5.5;
+  float Ki = 1.1;
+  float Kd = 1.5;
+  double output = 0;
+  PID myPID(&input, &output, &setpoint, Kp, Ki, Kd, DIRECT);
+  myPID.SetSampleTime(sample_time);
+  myPID.SetMode(AUTOMATIC);
+  myPID.Compute();
+
+  // Run PID a few times to have output in the middle of the range
+  for (int i = 0; i < 10; i++) {
+    Hal.delay(sample_time);
+    myPID.Compute();
+  }
+
+  double last_output = output;
+
+  // turn PID OFF
+  myPID.SetMode(MANUAL);
+
+  // Run PID a few times and check output doesn't change
+  for (int i = 0; i < 10; i++) {
+    Hal.delay(sample_time);
+    myPID.Compute();
+    EXPECT_EQ(output, last_output);
+  }
+
+  // change PID gains to make output easier to predict (this also tests
+  // SetTunings function while we are at it)
+  Kp = 0;
+  Ki = 1;
+  Kd = 0;
+  myPID.SetTunings(Kp, Ki, Kd);
+  // turn PID back ON
+  myPID.SetMode(AUTOMATIC);
+
+  double integral = (setpoint - input) * (sample_time / 1000.0);
+
+  myPID.Compute();
+
+  EXPECT_NEAR(output, last_output + integral * Ki, OUTPUT_TOLERANCE);
+}
```

Comment:
I think all of these tests are of the form "give the PID a fixed sequence of (time, input) pairs and verify what it does on each one", WDYT about refactoring them to actually look like this? Not saying you have to do this in this PR, or even that _you_ have to do this - but do you think this is a good idea in general?

Reply:
Yes! this is a very good idea, which I tried (and failed) to do at some point during my tests, and finally didn't do as I thought it wasn't worth spending too much time on.